guidedremediation/internal/manifest/python/poetry.go

```diff
@@ -0,0 +1,230 @@
+		return nil, fmt.Errorf("failed to unmarshal pyproject.toml: %w", err)
+	}
+
+	allReqs := []resolve.RequirementVersion{}
+	groups := make(map[manifest.RequirementKey][]string)
+
+	// Dependencies
+	allReqs = append(allReqs, parseDependencies(proj.Project.Dependencies, false)...)
+
+	// Optional dependencies
+	for groupName, deps := range proj.Project.OptionalDependencies {
+		groupReqs := parseDependencies(deps, true)
+		allReqs = append(allReqs, groupReqs...)
+		for _, r := range groupReqs {
+			key := manifest.RequirementKey(r.VersionKey.PackageKey)
+			groups[key] = append(groups[key], groupName)
+		}
+	}
+
+	return &pythonManifest{
+		filePath: path,
+		root: resolve.Version{
+			VersionKey: resolve.VersionKey{
+				PackageKey: resolve.PackageKey{
+					System: resolve.PyPI,
+					Name:   proj.Project.Name,
+				},
+				VersionType: resolve.Concrete,
+				Version:     proj.Project.Version,
+			},
+		},
+		requirements: allReqs,
+		groups:       groups,
+	}, nil
+}
+
+// Write writes the manifest after applying the patches to outputPath.
+func (r poetryReadWriter) Write(original manifest.Manifest, fsys scalibrfs.FS, patches []result.Patch, outputPath string) error {
+	return write(fsys, original.FilePath(), outputPath, patches, updatePyproject)
+}
+
+// updatePyproject takes an io.Reader representing the pyproject.toml file
+// and a map of package names to their new version constraints, returns the
+// file with the updated requirements as a string.
+func updatePyproject(reader io.Reader, requirements map[string][]VersionConstraint) (string, error) {
+	data, err := io.ReadAll(reader)
+	if err != nil {
+		return "", fmt.Errorf("error reading requirements: %w", err)
+	}
+
+	var sb strings.Builder
+	inProjectSection := false
+	inOptionalDepsSection := false
+	inDepArray := false
+	for _, line := range strings.SplitAfter(string(data), "\n") {
+		trimmedLine := strings.TrimSpace(line)
+
+		// Update current section.
+		if strings.HasPrefix(trimmedLine, "[") && strings.HasSuffix(trimmedLine, "]") {
+			inProjectSection = trimmedLine == "[project]"
+			inOptionalDepsSection = trimmedLine == "[project.optional-dependencies]"
+			inDepArray = false // Reset when section changes.
+		}
+
+		// Check if we are entering a dependency array.
+		if !inDepArray {
+			// `dependencies` in `[project]` or any key in `[project.optional-dependencies]`.
+			if (inProjectSection && strings.HasPrefix(trimmedLine, "dependencies =")) ||
+				(inOptionalDepsSection && strings.Contains(trimmedLine, "=")) {
+				if strings.Contains(line, "[") {
+					inDepArray = true
+				}
+			}
+		}
+
+		if !inDepArray {
+			// We don't need to touch non-dependency lines.
+			sb.WriteString(line)
+			continue
+		}
+
+		start := strings.Index(line, `"`)
+		if start < 0 {
+			// Requirement not found
+			sb.WriteString(line)
+			continue
+		}
+
+		end := strings.Index(line[start+1:], `"`)
+		if end < 0 {
+			// Closing quote is not found
+			sb.WriteString(line)
```

Comment:
nit: the `pyproject.toml` was probably invalid in this case - it's probably worth logging.